### PR TITLE
Exclude pointer assign from bitwise assign

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Fixes https://github.com/github/codeql-coding-standards/issues/389.

These queries were erroneously reporting pointer assign as a bitwise assign operator.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `M5-0-20`
     - `M5-0-21`
     - `RULE-10-1`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/common/src/codingstandards/cpp/Bitwise.qll [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * included.
+ */
+class AssignBitwiseOperationFixed extends AssignBitwiseOperation {
+  AssignBitwiseOperationFixed() {
+    // exclude += and -= on pointers, which seem to be erroneously included
+    // in the database schema
+    not this instanceof AssignPointerAddExpr and
+    not this instanceof AssignPointerSubExpr
+  }
+}
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/Bitwise.qll:18]:
```suggestion
/**
 * A library for addressing issues in bitwise operator modelling in our database schema.
 */

import cpp as cpp

module Bitwise {
  /**
   * A binary bitwise assign operation, excluding += and -= on pointers, which seem to be erroneously
   * included.
   */
  class AssignBitwiseOperation extends cpp::AssignBitwiseOperation {
    AssignBitwiseOperation() {
      // exclude += and -= on pointers, which seem to be erroneously included
      // in the database schema
      not this instanceof cpp::AssignPointerAddExpr and
      not this instanceof cpp::AssignPointerSubExpr
    }
  }
}
```

The suffix `Fixed` is ambiguous so I would like to suggest to have our own complete version and document the reasons for having our own version.

File: cpp/autosar/src/rules/M5-8-1/RightBitShiftOperandIsNegativeOrTooWide.ql [hunk 2]
```diff
@@ -33,7 +34,7 @@ class ShiftOperation extends Operation {
       rightOperand = o.getRightOperand()
     )
     or
-    exists(AssignBitwiseOperation o | this = o |
+    exists(AssignBitwiseOperationFixed o | this = o |
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M5-8-1/RightBitShiftOperandIsNegativeOrTooWide.ql:37]:
```suggestion
    exists(Bitwise::AssignBitwiseOperation o | this = o |
```

If we choose to rename our own version, we need to apply this change.

File: cpp/autosar/src/rules/M5-0-21/BitwiseOperatorAppliedToSignedTypes.ql [hunk 3]
```diff
@@ -17,14 +17,15 @@
...
+import codingstandards.cpp.Bitwise
 
 from Operation o, VariableAccess va
 where
   not isExcluded(o, ExpressionsPackage::bitwiseOperatorAppliedToSignedTypesQuery()) and
   (
     o instanceof UnaryBitwiseOperation or
     o instanceof BinaryBitwiseOperation or
-    o instanceof AssignBitwiseOperation
+    o instanceof AssignBitwiseOperationFixed
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M5-0-21/BitwiseOperatorAppliedToSignedTypes.ql:28]:
```suggestion
    o instanceof Bitwise::AssignBitwiseOperation
```

If we choose to rename our class, we need to apply this change

File: cpp/autosar/src/rules/M5-0-20/BitwiseOperatorOperandsHaveDifferentUnderlyingType.ql [hunk 4]
```diff
@@ -16,13 +16,17 @@
...
 import codingstandards.cpp.autosar
+import codingstandards.cpp.Bitwise
 
 predicate isBinaryBitwiseOperation(Operation o, VariableAccess l, VariableAccess r) {
   exists(BinaryBitwiseOperation bbo | bbo = o |
     l = bbo.getLeftOperand() and r = bbo.getRightOperand()
   )
   or
-  exists(AssignBitwiseOperation abo | abo = o | l = abo.getLValue() and r = abo.getRValue())
+  exists(AssignBitwiseOperationFixed abo | abo = o |
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M5-0-20/BitwiseOperatorOperandsHaveDifferentUnderlyingType.ql:26]:
```suggestion
  exists(Bitwise::AssignBitwiseOperation abo | abo = o |
```

If we choose to rename our class, we need to apply this change

File: c/misra/src/rules/RULE-10-1/OperandsOfAnInappropriateEssentialType.ql [hunk 5]
```diff
@@ -177,7 +178,7 @@ predicate isInappropriateEssentialType(
     child =
       [
         operator.(BinaryBitwiseOperation).getAnOperand(),
-        operator.(AssignBitwiseOperation).getAnOperand()
+        operator.(AssignBitwiseOperationFixed).getAnOperand()
```
Comment (rvermeulen) [inline c/misra/src/rules/RULE-10-1/OperandsOfAnInappropriateEssentialType.ql:181]:
```suggestion
        operator.(Bitwise::AssignBitwiseOperation).getAnOperand()
```

If we choose to rename our class, we need to apply this change

File: cpp/common/src/codingstandards/cpp/Bitwise.qll (hunk 1 above)
Comment (lcartey) [reply cpp/common/src/codingstandards/cpp/Bitwise.qll]: Agreed and changed.

**Comment by lcartey** [comment]
@rvermeulen all branches merged, requesting re-review.

**Review by rvermeulen — APPROVED** [review]
Looks good.

**Pull request merged**